File: packages/sdk-python/src/codewarden/transport.py
"""CodeWarden Transport - HTTP transport layer."""

from __future__ import annotations

import atexit
import logging
import queue
import threading
from typing import Any
from urllib.parse import urlparse

import httpx

from codewarden.exceptions import ConfigurationError
from codewarden.types import Event

logger = logging.getLogger(__name__)


def parse_dsn(dsn: str) -> tuple[str, str]:
    """
    Parse a CodeWarden DSN into base URL and API key.

    DSN format: https://API_KEY@host/path or just https://host/path with separate API key

    Examples:
        - https://cw_live_abc123@api.codewarden.io/v1/telemetry
        - https://cw_live_abc123@localhost:8000/v1/telemetry
        - https://api.codewarden.io/v1/telemetry (API key in URL or separate)

    Args:
        dsn: The Data Source Name string

    Returns:
        Tuple of (base_url, api_key)

    Raises:
        ConfigurationError: If DSN is invalid
    """
    if not dsn:
        raise ConfigurationError("DSN cannot be empty")

    parsed = urlparse(dsn)

    # Extract API key from URL if present (username part)
    api_key = parsed.username or ""

    # Build base URL without credentials
    if parsed.username:
        # Rebuild URL without the username (API key)
        netloc = parsed.hostname or ""
        if parsed.port:
            netloc = f"{netloc}:{parsed.port}"
        base_url = f"{parsed.scheme}://{netloc}"
    else:
        base_url = f"{parsed.scheme}://{parsed.netloc}"

    return base_url, api_key


class Transport:
    """HTTP transport for sending events to CodeWarden."""

    def __init__(
        self,
        dsn: str,
        *,
        max_queue_size: int = 100,
        batch_size: int = 10,
        flush_interval: float = 5.0,
        timeout: float = 30.0,
        max_retries: int = 3,
        debug: bool = False,
    ) -> None:
        """
        Initialize the transport.

        Args:
            dsn: Data Source Name (e.g., https://API_KEY@api.codewarden.io)
            max_queue_size: Maximum events to queue
            batch_size: Events per batch
            flush_interval: Seconds between flushes
            timeout: HTTP timeout in seconds
            max_retries: Maximum retry attempts
            debug: Enable debug logging
        """
        self._base_url, self._api_key = parse_dsn(dsn)
        self._batch_size = batch_size
        self._flush_interval = flush_interval
        self._timeout = timeout
        self._max_retries = max_retries
        self._debug = debug

        self._queue: queue.Queue[Event] = queue.Queue(maxsize=max_queue_size)
        self._shutdown = threading.Event()
        self._client = httpx.Client(timeout=timeout)

        # Start background worker
        self._worker = threading.Thread(target=self._worker_loop, daemon=True)
        self._worker.start()

        # Register cleanup on exit
        atexit.register(self.close)

    def send(self, event: Event) -> None:
        """
        Queue an event for sending.

        Args:
            event: The event to send
        """
        try:
            self._queue.put_nowait(event)
        except queue.Full:
            if self._debug:
                logger.warning("Event queue full, dropping event")

    def flush(self) -> None:
        """Flush all pending events."""
        events: list[Event] = []
        while True:
            try:
                event = self._queue.get_nowait()
                events.append(event)
            except queue.Empty:
                break

        for event in events:
            self._send_event(event)

    def close(self) -> None:
        """Close the transport and flush pending events."""
        self._shutdown.set()
        self.flush()
        self._client.close()

    def _worker_loop(self) -> None:
        """Background worker that sends events."""
        while not self._shutdown.is_set():
            try:
                # Wait for event with timeout
                event = self._queue.get(timeout=self._flush_interval)
                self._send_event(event)
            except queue.Empty:
                continue

    def _transform_event_to_payload(self, event: Event) -> dict[str, Any]:
        """
        Transform an SDK Event to the API's TelemetryPayload format.

        API expects:
        {
            "source": "backend-fastapi",
            "type": "error",
            "severity": "high",
            "environment": "production",
            "payload": { ... event data ... },
            "timestamp": "...",
            "trace_id": "..."
        }
        """
        # Map SDK level to API type/severity
        level = event.get("level", "info")
        event_type = "error" if level in ("error", "warning") else "info"
        severity_map = {
            "error": "high",
            "warning": "medium",
            "info": "low",
            "debug": "info",
        }
        severity = severity_map.get(level, "medium")

        # Build payload with exception details
        payload: dict[str, Any] = {
            "message": event.get("message"),
        }

        exception = event.get("exception")
        if exception:
            payload["error_type"] = exception.get("type")
            payload["error_message"] = exception.get("value")

            # Extract file and line from stack trace
            stacktrace = exception.get("stacktrace", [])
            if stacktrace:
                # Use the last frame (where the error occurred)
                last_frame = stacktrace[-1]
                payload["file"] = last_frame.get("filename")
                payload["line"] = last_frame.get("lineno")

                # Build full stack trace string
                trace_lines = []
                for frame in stacktrace:
                    line = f"  File \"{frame.get('filename')}\", line {frame.get('lineno')}, in {frame.get('function')}"
                    trace_lines.append(line)
                    if frame.get("context_line"):
                        trace_lines.append(f"    {frame.get('context_line')}")
                payload["stack_trace"] = "\n".join(trace_lines)

        # Add context data
        context = event.get("context", {})
        if context:
            payload["context"] = context

        # Add tags and extra data
        if event.get("tags"):
            payload["tags"] = event["tags"]
        if event.get("extra"):
            payload["extra"] = event["extra"]

        return {
            "source": "sdk-python",
            "type": event_type,
            "severity": severity,
            "environment": event.get("environment", "production"),
            "payload": payload,
            "timestamp": event.get("timestamp"),
            "trace_id": event.get("event_id"),  # Use event_id as trace_id for correlation
        }

    def _send_event(self, event: Event) -> None:
        """Send a single event with retry logic."""
        payload = self._transform_event_to_payload(event)
        endpoint = f"{self._base_url}/v1/telemetry"

        headers = {
            "Content-Type": "application/json",
        }

        # Add API key authorization if present
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"

        for attempt in range(self._max_retries):
            try:
                response = self._client.post(
                    endpoint,
                    json=payload,
                    headers=headers,
                )
                response.raise_for_status()
                if self._debug:
                    logger.debug(f"Sent event {event.get('event_id')} successfully")
                return
            except httpx.HTTPStatusError as e:
                if e.response.status_code < 500:
                    # Client error, don't retry
                    logger.error(f"Failed to send event: {e.response.status_code} - {e.response.text}")
                    return
                # Server error, retry
                if self._debug:
                    logger.warning(f"Server error (attempt {attempt + 1}): {e}")
            except httpx.RequestError as e:
                if self._debug:
                    logger.warning(f"Request error (attempt {attempt + 1}): {e}")

        logger.error(f"Failed to send event {event.get('event_id')} after {self._max_retries} attempts")

File: packages/sdk-js/src/transport.ts
/**
 * CodeWarden Transport - HTTP transport layer
 */

import type { Event } from './types';

interface TransportOptions {
  maxQueueSize?: number;
  flushInterval?: number;
  timeout?: number;
  maxRetries?: number;
  debug?: boolean;
}

interface TelemetryPayload {
  source: string;
  type: string;
  severity: string;
  environment: string;
  payload: Record<string, unknown>;
  timestamp?: string;
  trace_id?: string;
}

/**
 * Parse a CodeWarden DSN into base URL and API key.
 *
 * DSN format: https://API_KEY@host or https://API_KEY@host:port
 *
 * @example
 * parseDsn('https://cw_live_abc123@api.codewarden.io')
 * // Returns: { baseUrl: 'https://api.codewarden.io', apiKey: 'cw_live_abc123' }
 */
function parseDsn(dsn: string): { baseUrl: string; apiKey: string } {
  if (!dsn) {
    throw new Error('CodeWarden DSN cannot be empty');
  }

  try {
    const url = new URL(dsn);
    const apiKey = url.username || '';

    // Rebuild URL without credentials
    const baseUrl = `${url.protocol}//${url.host}`;

    return { baseUrl, apiKey };
  } catch {
    throw new Error(`Invalid CodeWarden DSN format: ${dsn}`);
  }
}

export class Transport {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly options: Required<Omit<TransportOptions, 'batchSize'>>;
  private queue: Event[] = [];
  private flushTimer: ReturnType<typeof setTimeout> | null = null;

  constructor(dsn: string, options: TransportOptions = {}) {
    const { baseUrl, apiKey } = parseDsn(dsn);
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;

    this.options = {
      maxQueueSize: options.maxQueueSize ?? 100,
      flushInterval: options.flushInterval ?? 5000,
      timeout: options.timeout ?? 30000,
      maxRetries: options.maxRetries ?? 3,
      debug: options.debug ?? false,
    };

    this.startFlushTimer();
  }

  /**
   * Queue an event for sending.
   */
  send(event: Event): void {
    if (this.queue.length >= this.options.maxQueueSize) {
      if (this.options.debug) {
        console.warn('[CodeWarden] Event queue full, dropping event');
      }
      return;
    }

    this.queue.push(event);
    // Send immediately instead of batching
    void this.flush();
  }

  /**
   * Flush all pending events.
   */
  async flush(): Promise<void> {
    while (this.queue.length > 0) {
      const event = this.queue.shift();
      if (event) {
        await this.sendEvent(event);
      }
    }
  }

  /**
   * Transform SDK Event to API TelemetryPayload format.
   */
  private transformEventToPayload(event: Event): TelemetryPayload {
    // Map SDK level to API type/severity
    const level = event.level;
    const eventType = level === 'error' || level === 'warning' ? 'error' : 'info';
    const severityMap: Record<string, string> = {
      error: 'high',
      warning: 'medium',
      info: 'low',
      debug: 'info',
    };
    const severity = severityMap[level] || 'medium';

    // Build payload with exception details
    const payload: Record<string, unknown> = {
      message: event.message,
    };

    const exception = event.exception;
    if (exception) {
      payload.error_type = exception.type;
      payload.error_message = exception.value;

      // Extract file and line from stack trace
      const stacktrace = exception.stacktrace || [];
      if (stacktrace.length > 0) {
        const lastFrame = stacktrace[stacktrace.length - 1];
        payload.file = lastFrame.filename;
        payload.line = lastFrame.lineno;

        // Build full stack trace string
        const traceLines: string[] = [];
        for (const frame of stacktrace) {
          traceLines.push(
            `  at ${frame.function} (${frame.filename}:${frame.lineno}:${frame.colno})`
          );
          if (frame.contextLine) {
            traceLines.push(`    ${frame.contextLine}`);
          }
        }
        payload.stack_trace = traceLines.join('\n');
      }
    }

    // Add context data
    if (event.context && Object.keys(event.context).length > 0) {
      payload.context = event.context;
    }

    // Add tags and extra data
    if (event.tags && Object.keys(event.tags).length > 0) {
      payload.tags = event.tags;
    }
    if (event.extra && Object.keys(event.extra).length > 0) {
      payload.extra = event.extra;
    }

    return {
      source: 'sdk-js',
      type: eventType,
      severity,
      environment: event.environment,
      payload,
      timestamp: event.timestamp,
      trace_id: event.eventId,
    };
  }

  /**
   * Send a single event with retry logic.
   */
  private async sendEvent(event: Event): Promise<void> {
    const payload = this.transformEventToPayload(event);
    const endpoint = `${this.baseUrl}/v1/telemetry`;

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    // Add API key authorization if present
    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    for (let attempt = 0; attempt < this.options.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.options.timeout);

        const response = await fetch(endpoint, {
          method: 'POST',
          headers,
          body: JSON.stringify(payload),
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (response.ok) {
          if (this.options.debug) {
            console.debug(`[CodeWarden] Sent event ${event.eventId} successfully`);
          }
          return;
        }

        if (response.status < 500) {
          // Client error, don't retry
          const text = await response.text();
          console.error(`[CodeWarden] Failed to send event: ${response.status} - ${text}`);
          return;
        }

        // Server error, retry
        if (this.options.debug) {
          console.warn(`[CodeWarden] Server error (attempt ${attempt + 1}): ${response.status}`);
        }
      } catch (error) {
        if (this.options.debug) {
          console.warn(`[CodeWarden] Request error (attempt ${attempt + 1}):`, error);
        }
      }

      // Wait before retry with exponential backoff
      await new Promise((resolve) => setTimeout(resolve, Math.pow(2, attempt) * 1000));
    }

    console.error(
      `[CodeWarden] Failed to send event ${event.eventId} after ${this.options.maxRetries} attempts`
    );
  }

  private startFlushTimer(): void {
    this.flushTimer = setInterval(() => {
      void this.flush();
    }, this.options.flushInterval);
  }

  /**
   * Stop the transport.
   */
  stop(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
  }
}
